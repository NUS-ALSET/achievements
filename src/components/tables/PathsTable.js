/**
 * @file PathsTable component module
 * @author Theodor Shaytanov <theodor.shaytanov@gmail.com>
 * @created 22.02.18
 */

import React from "react";
import PropTypes from "prop-types";

import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";
import Button from "@material-ui/core/Button";
import IconButton from "@material-ui/core/IconButton";

import SearchIcon from "@material-ui/icons/Search";
import EditIcon from "@material-ui/icons/Edit";

import { Link } from "react-router-dom";

import withStyles from "@material-ui/core/styles/withStyles";
import { pathDialogShow } from "../../containers/Paths/actions";
import { APP_SETTING } from "../../achievementsApp/config";

const styles = theme => ({
  link: {
    textDecoration: "none"
  },
  button: {
    margin: theme.spacing.unit
  }
});

class PathsTable extends React.PureComponent {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    dispatch: PropTypes.func.isRequired,
    owner: PropTypes.bool,
    paths: PropTypes.object.isRequired
  };

  onEditClick = path => this.props.dispatch(pathDialogShow(path));

  render() {
    const { classes, owner, paths } = this.props;

    return (
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>Path name</TableCell>
            {!owner && (
              <TableCell
                style={{
                  // eslint-disable-next-line no-magic-numbers
                  width: 150
                }}
              >
                Progress
              </TableCell>
            )}
            <TableCell
              style={{
                // eslint-disable-next-line no-magic-numbers
                width: APP_SETTING.isSuggesting ? 150 : 360
              }}
            >
              Actions
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {Object.keys(paths)
            .filter(id => paths[id])
            .map(id => ({ ...paths[id], id }))
            .map(path => (
              <TableRow key={path.id}>
                <TableCell>{path.name}</TableCell>
                {!owner && (
                  <TableCell>
                    {path.solutions !== undefined && path.totalActivities
                      ? `${path.solutions} of ${path.totalActivities}`
                      : ""}
                  </TableCell>
                )}
                {APP_SETTING.isSuggesting ? (
                  <TableCell>
                    <Link className={classes.link} to={`/paths/${path.id}`}>
                      <IconButton>
                        <SearchIcon />
                      </IconButton>
                    </Link>
                    {owner && (
                      <IconButton onClick={() => this.onEditClick(path)}>
                        <EditIcon />
                      </IconButton>
                    )}
                  </TableCell>
                ) : (
                  <TableCell>
                    <Link className={classes.link} to={`/paths/${path.id}`}>
                      <Button variant="raised">Open</Button>
                    </Link>
                    {owner && (
                      <Button
                        className={classes.button}
                        onClick={() => this.onEditClick(path)}
                        variant="raised"
                      >
                        Edit
                      </Button>
                    )}
                  </TableCell>
                )}
              </TableRow>
            ))}
        </TableBody>
      </Table>
    );
  }
}

export default withStyles(styles)(PathsTable);
