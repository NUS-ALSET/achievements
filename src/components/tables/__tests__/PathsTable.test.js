/* eslint-disable no-magic-numbers */
import React from "react";
import { createShallow } from "@material-ui/core/test-utils";
import sinon from "sinon";
import renderer from "react-test-renderer";
import TableCell from "@material-ui/core/TableCell";
import TableHead from "@material-ui/core/TableHead";
import { BrowserRouter as Router } from "react-router-dom";

import PathsTable from "../PathsTable";
import { TableBody, TableRow, IconButton } from "@material-ui/core";

describe("<PathsTable>", () => {
  let shallow;
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = sinon.spy();
    shallow = createShallow({
      dive: true
    });
  });

  it("should generate cells wihout throwing", () => {
    const wrapper = shallow(
      <PathsTable
        pathDialogShow={mockDispatch}
        paths={{
            "LEOF9q3Vdsd-gx1": {},
            "LEOF9dedq3V-gx1": {}
        }}
        viewCreatedTab={true}
      />
    );

    expect(wrapper.find(TableHead).find(TableCell).length).toEqual(2);
  });

  it("should generate rows without throwing", () => {
    const wrapper = shallow(
      <PathsTable
        pathDialogShow={mockDispatch}
        paths={{
            "LEOF9q3Vdsd-gx1": {},
            "LEOF9dedq3V-gx1": {}
        }}
        viewCreatedTab={true}
      />
    );
    expect(wrapper.find(TableBody).find(TableRow).length).toEqual(2);
  });

  it("should render row w/o edit button", () => {
    const wrapper = shallow(
      <PathsTable
        pathDialogShow={mockDispatch}
        paths={{
            "LEOF9q3Vdsd-gx1": {}
        }}
        viewCreatedTab={false}
      />
    );

    expect(wrapper.find(TableHead).find(TableCell).length).toEqual(3);
    expect(wrapper.find(IconButton).length).toEqual(1);
  });

  it("should render empty cell", () => {
    const wrapper = shallow(
      <PathsTable
        pathDialogShow={mockDispatch}
        paths={{}}
      />
    );
    expect(wrapper.find(TableBody).find(TableRow).length).toEqual(1);
  });

  it("should render cells for owner", () => {
    const wrapper = shallow(
      <PathsTable
        pathDialogShow={mockDispatch}
        paths={{}}
        viewCreatedTab={true}
      />
    );
    expect(wrapper.find(TableHead).find(TableRow).find(TableCell).length).toEqual(2);
  });

  it("renders correctly when there are no paths", () => {
    const tree = renderer.create(
      <PathsTable
        pathDialogShow={mockDispatch}
        paths={{}}
        viewCreatedTab={true}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly when there is one path", () => {
    const paths = {
      "LEOF9dedq3V-gx1": {}
    };
    const tree = renderer.create(
      <Router>
        <PathsTable
          pathDialogShow={mockDispatch}
          paths={paths}
          viewCreatedTab={true}
        />
      </Router>
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  it("renders correctly when there are multiple paths", () => {
    const paths = {
      "LEOF9dedq3V-gx1": {},
      "LEOF9dedgfV-gx3": {},
      "LEOF9debgbV-gx5": {}
      
    };
    const tree = renderer.create(
      <Router>
        <PathsTable
          pathDialogShow={mockDispatch}
          paths={paths}
          viewCreatedTab={true}
        />
      </Router>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
