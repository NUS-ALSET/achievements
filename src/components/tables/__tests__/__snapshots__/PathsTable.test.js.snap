// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PathsTable> renders correctly when there are multiple paths 1`] = `
<table
  className="MuiTable-root-2"
>
  <thead
    className="MuiTableHead-root-3"
  >
    <tr
      className="MuiTableRow-root-4 MuiTableRow-head-7"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-head-10"
        scope="col"
      >
        Path name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-head-10"
        scope="col"
        style={
          Object {
            "width": 150,
          }
        }
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-21"
  >
    <tr
      className="MuiTableRow-root-4 MuiTableRow-hover-6"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-body-11"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-body-11"
      >
        <a
          className="PathsTable-link-1"
          href="/paths/LEOF9dedq3V-gx1"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-28 MuiIconButton-root-22"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-27"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-31"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-40"
            />
          </button>
        </a>
        <button
          className="MuiButtonBase-root-28 MuiIconButton-root-22"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiIconButton-label-27"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-31"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-40"
          />
        </button>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root-4 MuiTableRow-hover-6"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-body-11"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-body-11"
      >
        <a
          className="PathsTable-link-1"
          href="/paths/LEOF9dedgfV-gx3"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-28 MuiIconButton-root-22"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-27"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-31"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-40"
            />
          </button>
        </a>
        <button
          className="MuiButtonBase-root-28 MuiIconButton-root-22"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiIconButton-label-27"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-31"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-40"
          />
        </button>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root-4 MuiTableRow-hover-6"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-body-11"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-body-11"
      >
        <a
          className="PathsTable-link-1"
          href="/paths/LEOF9debgbV-gx5"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-28 MuiIconButton-root-22"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-27"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-31"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-40"
            />
          </button>
        </a>
        <button
          className="MuiButtonBase-root-28 MuiIconButton-root-22"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiIconButton-label-27"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-31"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-40"
          />
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<PathsTable> renders correctly when there are no paths 1`] = `
<table
  className="MuiTable-root-2"
>
  <thead
    className="MuiTableHead-root-3"
  >
    <tr
      className="MuiTableRow-root-4 MuiTableRow-head-7"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-head-10"
        scope="col"
      >
        Path name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-head-10"
        scope="col"
        style={
          Object {
            "width": 150,
          }
        }
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-21"
  >
    <tr
      className="MuiTableRow-root-4"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-body-11"
        colSpan={3}
      >
        Empty
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<PathsTable> renders correctly when there is one path 1`] = `
<table
  className="MuiTable-root-2"
>
  <thead
    className="MuiTableHead-root-3"
  >
    <tr
      className="MuiTableRow-root-4 MuiTableRow-head-7"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-head-10"
        scope="col"
      >
        Path name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-head-10"
        scope="col"
        style={
          Object {
            "width": 150,
          }
        }
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-21"
  >
    <tr
      className="MuiTableRow-root-4 MuiTableRow-hover-6"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-body-11"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-9 MuiTableCell-body-11"
      >
        <a
          className="PathsTable-link-1"
          href="/paths/LEOF9dedq3V-gx1"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-28 MuiIconButton-root-22"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-27"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-31"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-40"
            />
          </button>
        </a>
        <button
          className="MuiButtonBase-root-28 MuiIconButton-root-22"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiIconButton-label-27"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-31"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-40"
          />
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
